misc-scripts/xref_mapping/XrefParser/Mim2GeneParser.pm

```diff
@@ -19,120 +19,281 @@ limitations under the License.
+      }
+      next RECORD;
     }
-    
-    if((!defined $mim_gene{$omim_id} ) and (!defined $mim_morbid{$omim_id} ) ){
-      $missed_omim++;
-      next;
+
+    if ( scalar @{ $line } != $EXPECTED_NUMBER_OF_COLUMNS ) {
+      croak ' Line ' . $csv->record_number()
+        . " of input file '${filename}' has an incorrect number of columns";
     }
 
-    if($type eq "gene" || $type eq 'gene/phenotype'){
-      if(defined($mim_gene{$omim_id})){
-	foreach my $ent_id (@{$entrez{$entrez_id}}){
-	  foreach my $mim_id (@{$mim_gene{$omim_id}}){
-	    $add_dependent_xref_sth->execute($ent_id, $mim_id);
-	  }
-	}
-	# $add_dependent_xref_sth->execute($entrez{$entrez_id}, $mim_gene{$omim_id});
-      }
-      else{
-	$diff_type++;
-	foreach my $ent_id (@{$entrez{$entrez_id}}){
-	  foreach my $mim_id (@{$mim_morbid{$omim_id}}){
-	    $add_dependent_xref_sth->execute($ent_id, $mim_id);
-	  }
-	}
-	# $add_dependent_xref_sth->execute($entrez{$entrez_id}, $mim_morbid{$omim_id});	
-      }
+    # Do not modify the contents of @{$line}, only the output - hence the /r.
+    my ( $omim_acc, $type, $entrez_id, $hgnc_symbol, $ensembl_id )
+      = map { s{\s+\z}{}rmsx } @{ $line };
+
+    $counters{'all_entries'}++;
+
+    # No point in doing anything if we have no matching MIM xref...
+    if ( ( !defined $mim_gene{$omim_acc} ) &&
+         ( !defined $mim_morbid{$omim_acc} ) )
+    {
+      $counters{'missed_omim'}++;
+      next RECORD;
     }
-    elsif($type eq "phenotype"){
-      if(defined($mim_morbid{$omim_id})){
-	foreach my $ent_id (@{$entrez{$entrez_id}}){
-	  foreach my $mim_id (@{$mim_morbid{$omim_id}}){
-	    $add_dependent_xref_sth->execute($ent_id, $mim_id);
-	  }
-	}
-	#	$add_dependent_xref_sth->execute($entrez{$entrez_id}, $mim_morbid{$omim_id});
-      }
-      else{
-	$diff_type++;
-	foreach my $ent_id (@{$entrez{$entrez_id}}){
-	  foreach my $mim_id (@{$mim_gene{$omim_id}}){
-	    $add_dependent_xref_sth->execute($ent_id, $mim_id);
-	  }
-	}
-	#	$add_dependent_xref_sth->execute($entrez{$entrez_id}, $mim_gene{$omim_id});	
-      }
+
+    # ...or no Ensembl ID or EntrezGene xref to match it to
+    # FIXME: this number might be underestimated because it doesn't
+    # check if Ensembl IDs from the input file actually exist
+    if ( ( ! $ensembl_id )
+         && ( ( ! $entrez_id ) || ( ! defined $entrez{$entrez_id} ) ) ) {
+      $counters{'missed_master'}++;
+      next RECORD;
     }
-    else{
-      print "WARNING unknown type $type\n";
-      next;
+
+    # An unknown type might indicate the change of input format,
+    # therefore make sure the user notices it. That said, do not
+    # bother we do not have an xref this entry would operate on anyway
+    # - which is why we only check this after the preceding two
+    # presence checks.
+    if ( ( $type ne 'gene')
+         && ( $type ne 'gene/phenotype' )
+         && ( $type ne 'predominantly phenotypes' )
+         && ( $type ne 'phenotype' ) ) {
+      croak "Unknown type $type";
```

Comment:
Could this error include a line/record number, or just the MIM ID? It would help with debugging.

Reply:
...or both, because why not.